File: vol1/ch11/LeetCode724.cpp
class Solution {
public:
    int numSubmatrixSumTarget(vector<vector<int>>& matrix, int target) {
        int m=matrix.size();
        int n=matrix[0].size();
        unordered_map<int,int> hmap;
        int ans=0;
        for(int i1=0;i1<m;i1++) {
            vector<int> colsum(n,0);
            for(int i2=i1;i2<m;i2++) {
                hmap[0]=1;						//一维前缀和数组(方式二)
                int cursum=0;
                for(int j=0;j<n;j++) {
                    colsum[j]+=matrix[i2][j];
                    cursum+=colsum[j];
                    ans+=hmap[cursum-target];
                    hmap[cursum]++;
                }
                hmap.clear();
            }
        }
        return ans;
    }
};

File: vol1/ch11/LeetCode724.py
class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        n=len(nums)
        presum=[0]*(n+1)
        presum[0]=0
        for i in range(1,n+1):
            presum[i]=presum[i-1]+nums[i-1]
        ans=n
        for i in range(0,n):
            if presum[i]==presum[n]-presum[i+1]:
                ans=i
                break
        if ans==n:return -1
        else: return ans
